### PR TITLE
[VL] Enable split preloading by default

### PR DESCRIPTION
**Pull request by zhli1142015**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

(Fixes: \#ISSUE-ID)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

File: shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala [hunk 1]
```diff
@@ -1211,7 +1211,7 @@ object GlutenConfig {
       .internal()
       .doc("The IO threads for connector split preloading")
       .intConf
-      .createWithDefault(0)
+      .createWithDefaultFunction(() => SQLConf.get.getConfString("spark.executor.cores", "1").toInt)
```
Comment (Yohahaha) [inline shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala:1214]: may I ask why use spark.executor.cores as default values? why not set 2 to align with Velox?

Comment (Yohahaha) [inline shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala:1214]: this thread executor will be used in DirectBufferedInput, not just for split preloading.

**Comment by Yohahaha** [comment]
I remember split preloading has memory issue you mentioned before, does that fixed?

**Comment by Yohahaha** [comment]
> I remember split preloading has memory issue you mentioned before, does that fixed?

I see the PR has merged in upstream Velox, thank you!

File: shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala (hunk 1 above)
Comment (zhli1142015) [reply shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala]: Updated the comments, thanks.

**Comment by zhli1142015** [comment]
cc @FelixYBW  and @zhouyuan , thanks.

**Comment by zhli1142015** [comment]
@zhouyuan  and @zhztheplayer , could you help check the memory leak failure? From the log I added and my understaning, this is related to the change https://github.com/apache/incubator-gluten/pull/5558/files#diff-1f31118af1f2bba25c508f4499fbe0304a65f1a681669ab555ce5ba03ed004c4.  The reservation is added twice (`pool->grow(0, bytes)` and `pool->grow(bytes, bytes)`) and only one is counted by child pool and can be released. 
BTW,  the memory leask looks not triggered by spilt preloading, but the `facebook::velox::cache::CoalescedLoad::loadOrFuture`.

image

File: shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala [hunk 2]
```diff
@@ -1230,9 +1230,10 @@ object GlutenConfig {
   val COLUMNAR_VELOX_CONNECTOR_IO_THREADS =
     buildStaticConf("spark.gluten.sql.columnar.backend.velox.IOThreads")
       .internal()
-      .doc("The IO threads for connector split preloading")
+      .doc("The Size of the IO thread pool in the Connector. This thread pool is used for split" +
+        " preloading and DirectBufferedInput.")
       .intConf
-      .createWithDefault(0)
+      .createWithDefaultFunction(() => SQLConf.get.getConfString("spark.executor.cores", "1").toInt)
```
Comment (ulysses-you) [inline shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala:1236]: Do you mean if the executor cores is 4, then there are at most 4 * 4 IO Threads at native side in one executor?

Comment (ulysses-you) [reply shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala:1236]: Or the IOThread is shared by all tasks in one executor ?

Comment (zhli1142015) [reply shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala:1236]: It's shared by all tasks in one exectuor.

**Comment by zhouyuan** [comment]
related:
https://github.com/apache/incubator-gluten/pull/3876

File: shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala (hunk 2 above)
Comment (jackylee-ch) [reply shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala:1236]: Does the IO Threads is related to the task num in one executor? The task num in one executor is determined by `spark.executor.cores`/`spark.task.cpus`, so we may get 8 IO Threads while there are only 4 tasks running in one executor.

**Review by Yohahaha — APPROVED** [review]
thank you! 

users need to be aware that this config can have a huge impact on performance.

File: shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala (hunk 2 above)
Comment (zhli1142015) [reply shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala]: Make sense, updated.

**Pull request merged**